### PR TITLE
feat: Add @cypress/puppeteer plugin

### LINKED ISSUE #28023: Create a puppeteer plugin
### What would you like?

Create a plugin that makes it easy to interact with the Cypress browser using Puppeteer. The plugin will require separate installation.

Message handlers are defined inside the Cypress config for utilizing Puppeteer like so:

```js
// cypress.config.ts
import { setup, retry } from '@cypress/puppeteer'
import { defineConfig } from 'cypress'
import type { Browser as PuppeteerBrowser, Page } from 'puppeteer-core'

export default defineConfig({
  e2e: {
    setupNodeEvents (on) {
      setup({
        on,
        onMessage: {
		  // 'testNewTab' can be any string and there can be as many uniquely-named message handlers as needed
          async testNewTab (browser: PuppeteerBrowser) {
            const page = await retry<Promise<Page>>(async () => {
              const pages = await browser.pages()
              const page = pages.find((page) => page.url().includes('new-tab.html'))

              if (!page) throw new Error('Could not find page in usage')

              return page
            })

            const paragraph = await page.waitForSelector('p')
            const paragraphText = await page.evaluate((el) => el!.textContent, paragraph)
            const button = await page.waitForSelector('#send')
            const buttonText = await page.evaluate((el) => el!.textContent, button)

            button!.dispose()
            paragraph!.dispose()

            await page.close()

            return { paragraphText, buttonText }
          },
        },
      })
    },
  },
})
```

Note that a generic retry function is provided as well.

The message handler is invoked in the spec by specifying the message via the custom `cy.puppeteer()` command:

```js
// spec.cy.ts
it('tests a second tab', () => {
    cy.visit('/cypress/fixtures/index.html')
    cy.get('button').click() // opens a new tab

    // this invokes the testNewTab message handler
    cy.puppeteer('testNewTab')
    .should('deep.equal', {
      paragraphText: 'This is the new tab',
      buttonText: 'Send message',
    })
})
```

The Cypress app itself should be updated to expect extra tabs and/or windows (any tabs/windows outside the main Cypress tab) being opened. 

### Why is this needed?

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Closes #28023

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/5565
- [N/A] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: npm/puppeteer/README.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+import { setup } from '@cypress/puppeteer/plugin'
+
+export default defineConfig({
+  e2e: {
+    setupNodeEvents (on) {
+      setup({
+        on,
+        onMessage: {
+          async switchToTabAndGetContent (browser) {
+            // ...
```
Comment (jennifer-shehane) [inline npm/puppeteer/README.md]: @chrisbreiding I would probably put an explicit comment here that says you need to now write Puppeteer code to execute what you want. I think someone might copy/paste this and assume it should do what it says. 🙂

File: npm/puppeteer/README.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    }
+  }
+})
+```
+
+## Examples
+
+These examples can be found and run in the [Cypress tests of this package](./cypress) with this project's [cypress.config.ts](./cypress.config.ts).
+
+While these example use tabs, they could just as easily apply to windows. Tabs and windows are essentially the same things as far as Puppeteer is concerned and encapsuled by instances of the [Page class](https://pptr.dev/api/puppeteer.page/).
```
Comment (jennifer-shehane) [inline npm/puppeteer/README.md]:
```suggestion
While these examples use tabs, they could just as easily apply to windows. Tabs and windows are essentially the same things as far as Puppeteer is concerned and encapsulated by instances of the [Page class](https://pptr.dev/api/puppeteer.page/).
```

File: npm/puppeteer/test/unit/retry.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  })
+
+  it('fails if function does not pass before timeout', async () => {
+    const fn = sinon.stub().callsFake(() => {
+      throw new Error('fail')
+    })
+
+    await expect(
+      retry(fn, { timeout: 5, delayBetweenTries: 1 }),
+    ).to.be.rejectedWith('ailed retrying after 5ms: fail')
```
Comment (jennifer-shehane) [inline npm/puppeteer/test/unit/retry.spec.ts]:
```suggestion
    ).to.be.rejectedWith('Failed retrying after 5ms: fail')
```

**Review by jennifer-shehane — APPROVED** [review]
Very clean and great documentation. Just a few comments.

File: npm/puppeteer/README.md (hunk 1 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Good idea. Addressed in https://github.com/cypress-io/cypress/pull/28370/commits/1d8356a92cf050de9c63813c72115bfcb160418e

File: npm/puppeteer/README.md (hunk 2 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Addressed in https://github.com/cypress-io/cypress/pull/28370/commits/1d8356a92cf050de9c63813c72115bfcb160418e

File: npm/puppeteer/test/unit/retry.spec.ts (hunk 3 above)
Comment (chrisbreiding) [reply npm/puppeteer/test/unit/retry.spec.ts]: Addressed in https://github.com/cypress-io/cypress/pull/28370/commits/1d8356a92cf050de9c63813c72115bfcb160418e

File: npm/puppeteer/README.md [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+## npm
+
+```sh
+npm install --save-dev @cypress/puppeteer
+```
+
+## yarn
+
+```sh
+yarn add @cypress/puppeteer
```
Comment (mschile) [inline npm/puppeteer/README.md]:
```suggestion
yarn add @cypress/puppeteer --dev
```

File: npm/puppeteer/README.md (hunk 1 above)
Comment (mschile) [reply npm/puppeteer/README.md]: Was this addressed? I don't see an explicit comment?

Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Ah good catch, I had put the comment [here](https://github.com/cypress-io/cypress/commit/1d8356a92cf050de9c63813c72115bfcb160418e#diff-bad8df9ac3ed2d72ad00ba7b71b5a9a6f51364a91eaf05c7ba1c9b1895fab279R225) but it's a good idea to add it in this spot as well.

File: npm/puppeteer/README.md [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+setup(options)
+```
+
+#### Options
+
+- `on` _required_: The `on` event registration function provided by `setupNodeEvents`
+- `onMessage` _required_: An object with string keys and function values (see more details [below](#onmessage))
+- `puppeteer` _optional_: The `puppeteer` library imported from `puppeteer-core`, overriding the default version of `puppeteer-core` used by this plugin
+
+#### onMessage
```
Comment (mschile) [inline npm/puppeteer/README.md]: Do we want to explicitly call out that you can't run cypress commands in the message handlers?

File: npm/puppeteer/README.md [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+{
+  "compilerOptions": {
+    "types": ["cypress", "@cypress/puppeteer"]
+  }
+}
+```
+
+## Compatibility
+
+`@cypress/puppeteer` requires Cypress version 13.6.0 or greater.
```
Comment (mschile) [inline npm/puppeteer/README.md:43]: Do we want to add a `peerDependency` in the `package.json` for this requirement?

File: npm/puppeteer/README.md (hunk 1 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Actually addressed this time [here](https://github.com/cypress-io/cypress/pull/28370/commits/beb85942ff50c82e10405d142cc6b0bc820580c6#diff-bad8df9ac3ed2d72ad00ba7b71b5a9a6f51364a91eaf05c7ba1c9b1895fab279R65)

File: npm/puppeteer/README.md (hunk 4 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Addressed [here](https://github.com/cypress-io/cypress/pull/28370/commits/beb85942ff50c82e10405d142cc6b0bc820580c6#diff-bad8df9ac3ed2d72ad00ba7b71b5a9a6f51364a91eaf05c7ba1c9b1895fab279R26)

File: npm/puppeteer/README.md (hunk 5 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md]: Yeah I like it. Addressed [here](https://github.com/cypress-io/cypress/pull/28370/commits/beb85942ff50c82e10405d142cc6b0bc820580c6#diff-bad8df9ac3ed2d72ad00ba7b71b5a9a6f51364a91eaf05c7ba1c9b1895fab279R113)

File: npm/puppeteer/README.md (hunk 6 above)
Comment (chrisbreiding) [reply npm/puppeteer/README.md:43]: Good idea. Addressed [here](https://github.com/cypress-io/cypress/pull/28370/commits/beb85942ff50c82e10405d142cc6b0bc820580c6#diff-dbb2ec8f0a23ac83b2225ec8d8459d75e8490d25b7f146ea51de157bbd9815f6R35-R37)

**Review by mschile — APPROVED** [review]

**Pull request merged**